File: cli/src/fp_cli/commands/apps.py
"""fp list, fp status, fp open, fp delete — app management commands."""

import webbrowser
from urllib.parse import urlparse

import typer
from rich.table import Table

from ..api.client import PlatformClient, PlatformError, resolve_app
from ..config import get_active_platform_or_exit
from ..console import console, err_console
from ..project import read_project, find_project_file


def _get_app_domain(platform_url: str) -> str:
    """Derive app domain from platform URL."""
    host = urlparse(platform_url).hostname or ""
    if host.startswith("platform."):
        return host[len("platform."):]
    return host


def _get_app_url(app_id: str, platform_url: str) -> str:
    return f"https://app-{app_id}.{_get_app_domain(platform_url)}"


def _resolve_app_name(name: str | None) -> str:
    """Resolve app name from argument or .fp.yaml."""
    if name:
        return name
    fp_file = find_project_file()
    if fp_file:
        import yaml

        data = yaml.safe_load(fp_file.read_text()) or {}
        if data.get("name"):
            return data["name"]
    err_console.print("[red]No app name provided and no .fp.yaml found.[/red]")
    raise typer.Exit(1)


def list_apps():
    """List all your apps."""
    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    try:
        apps = client.list_apps()
    except PlatformError as e:
        err_console.print(f"[red]Failed to list apps:[/red] {e.message}")
        raise typer.Exit(1)

    if not apps:
        console.print("[dim]No apps yet.[/dim] Run [bold]fp init[/bold] + [bold]fp deploy[/bold] to create one.")
        return

    table = Table(show_header=True, header_style="bold")
    table.add_column("Name")
    table.add_column("Status")
    table.add_column("URL")
    table.add_column("Last Deploy")

    for app in apps:
        status = app.get("status", "unknown")
        style = {
            "running": "green",
            "deploying": "yellow",
            "error": "red",
            "failed": "red",
        }.get(status, "dim")

        url = _get_app_url(app["app_id"], platform["url"])
        last_deploy = app.get("last_deploy_at") or app.get("created_at") or ""
        if last_deploy and "T" in last_deploy:
            last_deploy = last_deploy.split("T")[0]

        table.add_row(
            app["name"],
            f"[{style}]{status}[/{style}]",
            url,
            last_deploy,
        )

    console.print(table)


def status(name: str = typer.Argument(None, help="App name (uses .fp.yaml if omitted)")):
    """Show app status."""
    app_name = _resolve_app_name(name)
    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    app = resolve_app(client, app_name)
    app_url = _get_app_url(app["app_id"], platform["url"])

    try:
        ds = client.deploy_status(app["app_id"])
    except PlatformError:
        ds = {}

    app_status = ds.get("status") or app.get("status", "unknown")
    ready = ds.get("deployment_ready", False)

    style = "green" if app_status == "running" and ready else "yellow" if app_status == "deploying" else "red" if app_status in ("error", "failed") else "dim"

    console.print(f"  Name:   [bold]{app['name']}[/bold]")
    console.print(f"  Status: [{style}]{app_status}[/{style}]")
    console.print(f"  Ready:  {'yes' if ready else 'no'}")
    console.print(f"  URL:    {app_url}")
    console.print(f"  ID:     {app['app_id']}")
    if ds.get("last_error"):
        console.print(f"  Error:  [red]{ds['last_error']}[/red]")


def open_app(name: str = typer.Argument(None, help="App name (uses .fp.yaml if omitted)")):
    """Open app URL in browser."""
    app_name = _resolve_app_name(name)
    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    app = resolve_app(client, app_name)
    url = _get_app_url(app["app_id"], platform["url"])

    console.print(f"Opening [link={url}]{url}[/link]")
    webbrowser.open(url)


def delete(
    name: str = typer.Argument(..., help="App name to delete"),
    yes: bool = typer.Option(False, "--yes", "-y", help="Skip confirmation"),
):
    """Delete an app."""
    platform = get_active_platform_or_exit()
    client = PlatformClient(platform["url"], platform["token"])

    app = resolve_app(client, name)

    if not yes:
        confirm = typer.confirm(f"Delete app '{app['name']}'? This cannot be undone")
        if not confirm:
            raise typer.Abort()

    try:
        client.delete_app(app["app_id"])
    except PlatformError as e:
        err_console.print(f"[red]Delete failed:[/red] {e.message}")
        raise typer.Exit(1)

    console.print(f"Deleted [bold]{app['name']}[/bold]")

File: cli-go/cmd/apps.go
package cmd

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strings"

	"github.com/spf13/cobra"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/api"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/config"
	"github.com/thatcatxedo/fastapi-platform/cli-go/internal/project"
)

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "List all your apps",
	RunE:  runList,
}

var statusCmd = &cobra.Command{
	Use:   "status [name]",
	Short: "Show app status",
	Args:  cobra.MaximumNArgs(1),
	RunE:  runStatus,
}

var openCmd = &cobra.Command{
	Use:   "open [name]",
	Short: "Open app URL in browser",
	Args:  cobra.MaximumNArgs(1),
	RunE:  runOpen,
}

var deleteCmd = &cobra.Command{
	Use:   "delete [name]",
	Short: "Delete an app",
	Args:  cobra.ExactArgs(1),
	RunE:  runDelete,
}

var deleteYes bool

func init() {
	deleteCmd.Flags().BoolVarP(&deleteYes, "yes", "y", false, "Skip confirmation")
}

func requirePlatform() *config.Platform {
	p := config.GetActivePlatform()
	if p == nil {
		fmt.Fprintln(os.Stderr, "Not authenticated. Run 'fp auth login <platform-url>' first.")
		os.Exit(1)
	}
	return p
}

func resolveAppName(name string) string {
	if name != "" {
		return name
	}
	fp, ok := project.FindProjectFile("")
	if !ok {
		fmt.Fprintln(os.Stderr, "No app name provided and no .fp.yaml found.")
		os.Exit(1)
	}
	proj, err := project.Read(fp)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	if proj.Name == "" {
		fmt.Fprintln(os.Stderr, "No app name in .fp.yaml")
		os.Exit(1)
	}
	return proj.Name
}

func runList(cmd *cobra.Command, args []string) error {
	p := requirePlatform()
	client := api.NewClient(p.URL, p.Token)
	apps, err := client.ListApps()
	if err != nil {
		var pe *api.PlatformError
		if errors.As(err, &pe) {
			fmt.Fprintf(os.Stderr, "Failed to list apps: %s\n", pe.Message)
			os.Exit(1)
		}
		return err
	}
	if len(apps) == 0 {
		fmt.Println("No apps yet. Run 'fp init' + 'fp deploy' to create one.")
		return nil
	}

	// Table
	fmt.Printf("%-20s %-10s %-40s %-12s\n", "Name", "Status", "URL", "Last Deploy")
	fmt.Println(strings.Repeat("-", 85))
	for _, a := range apps {
		name, _ := a["name"].(string)
		status, _ := a["status"].(string)
		appID, _ := a["app_id"].(string)
		url := api.AppURL(p.URL, appID)
		lastDeploy, _ := a["last_deploy_at"].(string)
		if lastDeploy == "" {
			lastDeploy, _ = a["created_at"].(string)
		}
		if idx := strings.Index(lastDeploy, "T"); idx > 0 {
			lastDeploy = lastDeploy[:idx]
		}
		if len(url) > 38 {
			url = url[:35] + "..."
		}
		fmt.Printf("%-20s %-10s %-40s %-12s\n", name, status, url, lastDeploy)
	}
	return nil
}

func runStatus(cmd *cobra.Command, args []string) error {
	appName := resolveAppName("")
	if len(args) > 0 {
		appName = args[0]
	}
	p := requirePlatform()
	client := api.NewClient(p.URL, p.Token)
	app, err := api.ResolveApp(client, appName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	appID, _ := app["app_id"].(string)
	appURL := api.AppURL(p.URL, appID)

	ds, _ := client.DeployStatus(appID)
	appStatus, _ := ds["status"].(string)
	if appStatus == "" {
		appStatus, _ = app["status"].(string)
	}
	if appStatus == "" {
		appStatus = "unknown"
	}
	ready, _ := ds["deployment_ready"].(bool)

	style := "dim"
	if appStatus == "running" && ready {
		style = "green"
	} else if appStatus == "deploying" {
		style = "yellow"
	} else if appStatus == "error" || appStatus == "failed" {
		style = "red"
	}

	fmt.Printf("  Name:   %s\n", app["name"])
	fmt.Printf("  Status: %s\n", appStatus)
	fmt.Printf("  Ready:  %v\n", ready)
	fmt.Printf("  URL:    %s\n", appURL)
	fmt.Printf("  ID:     %s\n", appID)
	if lastErr, _ := ds["last_error"].(string); lastErr != "" {
		fmt.Printf("  Error:  %s\n", lastErr)
	}
	_ = style
	return nil
}

func runOpen(cmd *cobra.Command, args []string) error {
	appName := resolveAppName("")
	if len(args) > 0 {
		appName = args[0]
	}
	p := requirePlatform()
	client := api.NewClient(p.URL, p.Token)
	app, err := api.ResolveApp(client, appName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	appID, _ := app["app_id"].(string)
	appURL := api.AppURL(p.URL, appID)
	fmt.Printf("Opening %s\n", appURL)

	var openCmd *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		openCmd = exec.Command("open", appURL)
	case "windows":
		openCmd = exec.Command("rundll32", "url.dll,FileProtocolHandler", appURL)
	default:
		openCmd = exec.Command("xdg-open", appURL)
	}
	return openCmd.Start()
}

func runDelete(cmd *cobra.Command, args []string) error {
	appName := args[0]
	p := requirePlatform()
	client := api.NewClient(p.URL, p.Token)
	app, err := api.ResolveApp(client, appName)
	if err != nil {
		fmt.Fprintln(os.Stderr, err.Error())
		os.Exit(1)
	}
	if !deleteYes {
		fmt.Printf("Delete app '%s'? This cannot be undone [y/N]: ", app["name"])
		var confirm string
		fmt.Scanln(&confirm)
		if !strings.EqualFold(strings.TrimSpace(confirm), "y") {
			return nil
		}
	}
	appID, _ := app["app_id"].(string)
	if err := client.DeleteApp(appID); err != nil {
		var pe *api.PlatformError
		if errors.As(err, &pe) {
			fmt.Fprintf(os.Stderr, "Delete failed: %s\n", pe.Message)
			os.Exit(1)
		}
		return err
	}
	fmt.Printf("Deleted %s\n", app["name"])
	return nil
}
